numpy/core/numeric.py

```diff
@@ -458,10 +465,31 @@ class ndarray is returned.
     >>> np.asanyarray(a) is a
     True
 
+    Make sure the result is compatible with a specified dtype
+
+    >>> a = np.ones(2, dtype=np.int16)
+    >>> np.asarray(a, precision=np.float32)
+    array([ 1.,  1.], dtype=float32)
+    >>> a = np.ones(2, dtype=np.complex128)
+    >>> np.asarray(a, precision=np.float32)
+    array([ 1.+0.j,  1.+0.j])
+    >>> a = np.ones(2, dtype=np.complex64)
+    >>> np.asarray(a, precision=0.0)
+    array([ 1.+0.j,  1.+0.j], dtype=complex64)
+
     """
-    return array(a, dtype, copy=False, order=order)
+    if precision is not None and dtype is not None:
+        raise ValueError('cannot specify both dtype and precision')
+
+    ret = array(a, dtype, copy=False, order=order)
```

Comment:
We call `asarray` a lot, and most of the time `a` is already an array. It may be a worthwhile optimization to, if `dtype` is `None`, but `isinstance(a, ndarray)`, try getting a dtype from `result_type(a.dtype, precision)` *before* calling `array` on `a`. It would save us a copy e.g. when asking for floating point precision on an integer array.